metric_learn/mmc.py

```diff
@@ -350,10 +355,7 @@ def _D_constraint(self, neg_pairs, w):
       sum_deri2 / sum_dist - np.outer(sum_deri1, sum_deri1) / (sum_dist * sum_dist)
     )
 
-  def metric(self):
-    return self.A_
-
-  def transformer(self):
+  def transformer_from_metric(self, metric):
```

Comment:
I can't see why we need to override `transformer_from_metric` for MMC?

Reply:
Agreed
I tried to remove it but `metric_learn_test.TestMMC` fails, throwing
```pytb
E       numpy.linalg.linalg.LinAlgError: Matrix is not positive definite
```
And indeed, looking into this test with the overwritten `_transformer_from_metric` with eigendecomposition, after doing the eigendecomposition, 3 out of 4 eigenvalues are very close to zero, so this would mean the metric is semi-definite (as stated in the code comment).
Cholesky decomposition is defined for semidefinite matrices, but looking it up it appears that the scipy's routine does not work in such a case (https://mail.scipy.org/pipermail/scipy-user/2011-July/029927.html). So I guess I could deal with semidefinite case inside the general `_transformer_from_metric` with maybe a little function to test if the matrix invertible with `numpy.linalg.det
` ? And in this case we would do the eigenvalues decomposition ?

Reply:
I think an easy solution is to apply Cholesky to `metric` plus a very small factor of the identity matrix